src/legacy-appcache-behavior-import.js

```diff
@@ -0,0 +1,422 @@
+      return Promise.all(idsOfInactiveClients.map(id => {
+        return idb.get(constants.STORES.CLIENT_ID_TO_HASH, id).then(hash => {
+          return idb.delete(constants.STORES.CLIENT_ID_TO_HASH, id).then(() => hash);
+        });
+      }));
+    });
+  }
+
+  /**
+   * Fulfills with an array of all the hash ids that correspond to outdated
+   * manifest versions.
+   *
+   * @returns {Promise.<Array.<String>>}
+   */
+  function getHashesOfOlderVersions() {
+    return idb.getAll(constants.STORES.MANIFEST_URL_TO_CONTENTS).then(manifests => {
+      return manifests.map(versions => {
+        // versions.slice(0, -1) will give all the versions other than the
+        // last, or [] if there's aren't any older versions.
+        return versions.slice(0, -1).map(version => version.hash);
+      }).reduce((prev, curr) => {
+        // Flatten the array-of-arrays into an array.
+        return prev.concat(curr);
+      }, []);
+    });
+  }
+
+  /**
+   * Does the following:
+   * 1. Gets a list of all client ids associated with this service worker.
+   * 2. Calls cleanupClientIdAndHash() to remove the out of date client id
+   *    to hash associations.
+   * 3. Calls getHashesOfOlderVersions() to get a list of all the hashes
+   *    that correspond to out-of-date manifest versions.
+   * 4. If there's a match between an out of date hash and a hash that is no
+   *    longer being used by a client, then it deletes the corresponding cache.
+   *
+   * @returns {Promise.<T>}
+   */
+  function cleanupOldCaches() {
+    self.clients.matchAll().then(clients => {
+      return clients.map(client => client.id);
+    }).then(idsOfActiveClients => {
+      return cleanupClientIdAndHash(idsOfActiveClients);
+    }).then(hashesNotInUse => {
+      return getHashesOfOlderVersions().then(hashesOfOlderVersions => {
+        return hashesOfOlderVersions.filter(hashOfOlderVersion => {
+          return hashesNotInUse.includes(hashOfOlderVersion);
+        });
+      });
+    }).then(idsToDelete => {
+      log('deleting cache ids', idsToDelete);
+      return Promise.all(idsToDelete.map(cacheId => caches.delete(cacheId)));
+    });
+
+    // TODO: Delete the entry in the array stored in MANIFEST_URL_TO_CONTENT.
+  }
+
+  /**
+   * A wrapper on top of appCacheBehaviorForEvent() that handles rejections with
+   * a default of fetch().
+   *
+   * @param {FetchEvent} event
+   * @returns {Promise.<Response>}
+   */
+  global.legacyAppCacheBehavior = event => {
+    log('client id is', event.clientId);
+    return appCacheBehaviorForEvent(event).then(response => {
+      // If this is a navigation, clean up unused caches that correspond to old
+      // AppCache manifest versions which are no longer associated with an
+      // active client. This will be done asynchronously, and won't block the
+      // response from being returned to the onfetch handler.
+      if (event.request.mode === 'navigate') {
+        cleanupOldCaches();
```

Comment:
After in-depth reading of what this is doing, maybe you don't mind the process is suddenly interrupted although the grace time for the service worker should be enough to allow cleaning to complete.